src/main/JayZ.ts

```diff
@@ -1,49 +1,123 @@
-    const { encryptedItem } = await this.encryptor.encrypt({ item, fieldsToEncrypt })
-    return encryptedItem
+    const { dataKey, encryptedDataKey } = await this.getNextDataKey()
+    const { plaintextFields, nonce, encryptedFields } = this.encryptor.encrypt({
+      item,
+      fieldsToEncrypt,
+      dataKey
+    })
+
+    const __jayz__metadata: EncryptedItemMetadataV1 = {
+      encryptedDataKey,
+      nonce,
+      scheme: this.encryptor.scheme,
+      encryptedFields,
+      metadataVersion: MetadataVersion.V1
+    }
+
+    return {
+      ...plaintextFields,
+      __jayz__metadata
+    }
   }
 
-  async encryptItems<T, U extends keyof T>(
-    itemsToEncrypt: EncryptItemProps<T, U>[]
-  ): Promise<EncryptedJayZItem<T, U>[]> {
+  encryptItems<T, K extends keyof T>(
+    itemsToEncrypt: EncryptItemRequest<T, K>[]
+  ): Promise<EncryptedJayZItem<T, K>[]> {
     if (itemsToEncrypt.length === 0) {
-      return []
+      return Promise.resolve([])
     }
 
     const items = itemsToEncrypt.map((item) => this.encryptItem(item))
     return Promise.all(items)
   }
 
-  async decryptItem<T, U extends keyof T>(item: EncryptedJayZItem<T, U>): Promise<T> {
-    const { decryptedItem } = await this.encryptor.decrypt<T, U>({ item })
+  async decryptItem<T, K extends keyof T>(
+    itemToDecrypt: EncryptedJayZItem<T, K>
+  ): Promise<T> {
+    const {
+      nonce,
+      encryptedDataKey,
+    } = itemToDecrypt.__jayz__metadata
+
+    const encryptedFields = (itemToDecrypt.__jayz__metadata as EncryptedItemMetadataV1).encryptedFields
+    const encryptedFieldNames = (itemToDecrypt.__jayz__metadata as LegacyEncryptedItemMetadata<T, K>).encryptedFieldNames
+
+    const encryptedItem = { ...itemToDecrypt }
+    delete (encryptedItem as any).__jayz__metadata
+
+    const dataKey = await this.keyProvider.decryptDataKey(encryptedDataKey)
+    const { decryptedItem } = this.encryptor.decrypt<T, K>({
+      encryptedItem,
+      fieldsToDecrypt: encryptedFieldNames,
+      dataKey,
+      nonce,
+      encryptedFields
+    })
+
+    memzero(dataKey)
     return decryptedItem
   }
 
-  async decryptItems<T, U extends keyof T>(itemsToDecrypt: EncryptedJayZItem<T, U>[]): Promise<T[]> {
+  decryptItems<T, K extends keyof T>(
+    itemsToDecrypt: EncryptedJayZItem<T, K>[]
+  ): Promise<T[]> {
     if (itemsToDecrypt.length === 0) {
-      return []
+      return Promise.resolve([])
     }
 
     const itemPromises = itemsToDecrypt.map((item) => this.decryptItem(item))
     return Promise.all(itemPromises)
   }
+
+  private getNextDataKey(): Promise<DataKey> {
+    if (
+      this.currentDataKey !== undefined &&
+      this.currentDataKeyUsesRemaining > 0
+    ) {
+      this.currentDataKeyUsesRemaining -= 1
+      return this.currentDataKey
+    } else {
+      this.currentDataKey = this.keyProvider.generateDataKey()
+      this.currentDataKeyUsesRemaining = this.maxUsesPerDataKey - 1
+      return this.currentDataKey
+    }
+  }
```

Comment:
This looks to be inlined (with the goal of simplifying) from `LibsodiumKdfKeyProvider` which went away.